lib/rack/utf8_sanitizer.rb

```diff
@@ -167,8 +170,12 @@ def close
       end
     end
 
-    def sanitize_io(io, uri_encoded = false)
-      input = io.read
+    def sanitize_io(io, uri_encoded = false, content_length = nil)
+      input = if content_length
+        io.read(content_length)
```

Comment:
If there any risk to crash/block if the `CONTENT_LENGTH` sent by the client is negative?

Reply:
That is dependent on the underlying server. But given that Rails will read the request body by passing `content_length`, I believe it's already handled by servers.

https://github.com/rails/rails/blob/939fd3bc41bbb90194e4b93a001c803bdceb592c/actionpack/lib/action_dispatch/http/request.rb#L467C14-L470

Reply:
As I know, you can send any value for any header. Yeah, it'd be out of standards, but it's still possible, and someone can use it (for example, as HTTP 200 OK with "not found" error).

Reply:
@AlexWayfer sorry I'm not sure what point you are trying to make.

Yes clients can send garbage in `Content-Length`, but that's no reason not to use it.

Reply:
As I understood, we're not talking about using it or not, we're discussing negative (or other) values which potentially could crash `io.read` code.

BTW, I see no tests changes in this PR, and this is… dangerous.

Reply:
Servers do handle negative Content-Length already.

> I see no tests changes in this PR

This codepath is already tested.

Reply:
> Servers do handle negative Content-Length already.

Which servers? Any servers? Thin, Puma, Unicorn, Falcon — any of them? I don't think we can rely this way. If Rack would handle, on which we rely (and depend) — then OK.

> This codepath is already tested.

If you introduce changes to the code — tests should be added or changed too.

For example, the old code always did `io.read`. Now we have condition — great, now we have 2 cases and both should be tested, truthy and falsey branches. Especially that `io` can contain more than `content_length`, but only `content_length` would be read, not the whole `io`.

Reply:
> Any servers? Thin, Puma, Unicorn, Falcon — any of them?

Yes. `rack.input` in Puma is either `StringIO` or `Tempfile`, both of which raise `ArgumentError` on negative size like Ruby's `IO`.

Unicorn/Pitchfork both explicitly raise argument error on such case too to mimick Ruby IO.

Falcon returns empty string on negative length.

Thin's `rack.input` is a `StringIO`, so raises as well.

Now the Rack spec says it's invalid to read a negative length, so I guess we can check for it.

> If you introduce changes to the code — tests should be added or changed too.

Not if the change is functionally equivalent.